cinder/volume/drivers/vmware/vmdk.py

```diff
@@ -651,18 +646,102 @@ def _create_snapshot(self, snapshot):
         """
 
         volume = snapshot['volume']
-        if volume['status'] != 'available':
+        if volume['status'] == 'available':
+            backing = self.volumeops.get_backing(snapshot['volume_name'])
+            if not backing:
+                LOG.info(_LI("There is no backing, so will not create "
+                             "snapshot: %s."), snapshot['name'])
+                return
+            self.volumeops.create_snapshot(backing, snapshot['name'],
+                                           snapshot['display_description'])
+        elif volume['status'] == 'in-use':
+            # Get a reference to the shadow VM that is backing the volume
+            backing_ref = self.volumeops.get_backing(snapshot['volume_name'])
+
+            # The volume is in use -> it is attached
+            all_attachments = volume['volume_attachment']
+
+            # Pick the first, regardless of multi-attach
+            attachment = all_attachments[0]
+
+            # Find the vm by its uuid
+            instance_uuid = attachment['instance_uuid']
+            vm_ref = self.volumeops.get_vm_ref_from_vm_uuid(instance_uuid)
+
+            # Snapshot the VM
+            snapshotName = "tempVolumeSnapshot"
+            snapshot_ref = self.volumeops.create_snapshot(vm_ref, snapshotName, None)
+
+            # Clone from snapshot (so that we use linked mode)
+            cloned_vm_ref = self.volumeops.clone_vm(vm_ref, snapshot_ref)
+
+            # Determine the disk ?
+            backing_disks = self.volumeops._get_disk_devices(backing_ref)
+            if len(backing_disks) != 1:
+                err_msg = "Unexpected number of virtual disks on backing vm."
+                LOG.error(err_msg)
+                raise Exception(err_msg)
+            backing_disk_uuid = backing_disks[0].backing.uuid
+            backing_disk_file_name = backing_disks[0].backing.fileName
+
+            cloned_disks = self.volumeops._get_disk_devices(cloned_vm_ref)
+            for disk_device in cloned_disks:
+                if disk_device.backing.uuid == backing_disk_uuid:
+                    break # disk found
+            else:
+                err_msg = "Unable to find matching disk in cloned vm."
+                LOG.error(err_msg)
+                raise Exception(err_msg)
+
+            # detach the new disk from the cloned vm
+            LOG.debug("Detaching new disk from cloned vm {}.".format(cloned_vm_ref))
+            self.volumeops.detach_disk_from_backing(cloned_vm_ref, disk_device)
```

Comment:
Don't you need to remove the "tempVolumeSnapshot" snapshot from the cloned vm?
Otherwise, you'll increase add in line 731 a delta to "tempVolumeSnapshot", which is (hopefully) none to the disk-chain, no?